File: android/app/src/main/java/app/workfence/workforce/BiometricPlugin.java
package app.workfence.workforce;

import android.os.Build;

import androidx.annotation.NonNull;
import androidx.biometric.BiometricManager;
import androidx.biometric.BiometricPrompt;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * The phone's own face or fingerprint, asked for at the gate.
 *
 * This proves the phone is in the hands of the person who set it up. That is
 * a different question from the one the selfie answers — whether the face at
 * the boundary is the enrolled worker's — and it is asked first because it is
 * instant and because a phone handed to a mate fails it before a camera is
 * ever opened.
 *
 * It never replaces the selfie. A biometric says yes or no and leaves nothing
 * behind; attendance needs a record somebody can look at later, and that is
 * the photograph and the descriptor comparison.
 *
 * DEVICE_CREDENTIAL is deliberately allowed alongside. A fingerprint sensor
 * on a site is read through dust, cement and rain, and a worker locked out of
 * their own day by a wet thumb would simply stop using the app. The PIN they
 * set is the same proof of possession, slower.
 */
@CapacitorPlugin(name = "Biometric")
public class BiometricPlugin extends Plugin {

    /**
     * BIOMETRIC_WEAK rather than STRONG: face unlock on a great many Android
     * phones is classed weak, and requiring STRONG would report "no biometric"
     * to exactly the users who think they have one. Nothing is being
     * encrypted here — the answer is a yes or no about who is holding the
     * phone, and weak is the right bar for that.
     */
    private static final int AUTHENTICATORS =
            BiometricManager.Authenticators.BIOMETRIC_WEAK
                    | BiometricManager.Authenticators.DEVICE_CREDENTIAL;

    /**
     * What this device can do, asked before anything is shown.
     *
     * The web layer uses this to decide whether to put a step in front of
     * check-in at all. A phone with no sensor and no PIN must not be given a
     * gate it can never pass.
     */
    @PluginMethod
    public void isAvailable(PluginCall call) {
        JSObject out = new JSObject();
        try {
            BiometricManager manager = BiometricManager.from(getContext());
            int status = manager.canAuthenticate(AUTHENTICATORS);
            out.put("available", status == BiometricManager.BIOMETRIC_SUCCESS);
            out.put("reason", describe(status));
            // Whether a face is what they will be shown, so the prompt can say
            // the word the person expects rather than a generic one.
            out.put(
                    "face",
                    Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q
                            && getContext()
                                    .getPackageManager()
                                    .hasSystemFeature(android.content.pm.PackageManager.FEATURE_FACE));
        } catch (Exception e) {
            out.put("available", false);
            out.put("reason", "unavailable");
            out.put("face", false);
        }
        call.resolve(out);
    }

    private String describe(int status) {
        switch (status) {
            case BiometricManager.BIOMETRIC_SUCCESS:
                return "ready";
            case BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED:
                return "nothing enrolled";
            case BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE:
                return "no hardware";
            case BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE:
                return "hardware busy";
            default:
                return "unavailable";
        }
    }

    /**
     * Ask, and answer with what happened rather than only whether it worked.
     *
     * The caller has to tell a refusal from a broken sensor: somebody who
     * cancelled can try again, and somebody whose phone has no working sensor
     * should be let through to the selfie rather than held at a door that
     * will not open.
     */
    @PluginMethod
    public void verify(PluginCall call) {
        final String title = call.getString("title", "Confirm it's you");
        final String subtitle = call.getString("subtitle", "");

        FragmentActivity activity = (FragmentActivity) getActivity();
        if (activity == null) {
            call.resolve(result(false, "unavailable"));
            return;
        }

        BiometricPrompt prompt =
                new BiometricPrompt(
                        activity,
                        ContextCompat.getMainExecutor(getContext()),
                        new BiometricPrompt.AuthenticationCallback() {
                            @Override
                            public void onAuthenticationSucceeded(
                                    @NonNull BiometricPrompt.AuthenticationResult r) {
                                call.resolve(result(true, "ok"));
                            }

                            @Override
                            public void onAuthenticationError(int code, @NonNull CharSequence msg) {
                                // A cancel is a person saying no; the rest is a
                                // device saying it cannot. The web layer treats
                                // them differently and so must this.
                                boolean cancelled =
                                        code == BiometricPrompt.ERROR_USER_CANCELED
                                                || code == BiometricPrompt.ERROR_NEGATIVE_BUTTON
                                                || code == BiometricPrompt.ERROR_CANCELED;
                                call.resolve(result(false, cancelled ? "cancelled" : "failed"));
                            }

                            // Deliberately not resolving on a single bad read:
                            // the prompt stays up and lets them try again, which
                            // is what somebody with a dusty thumb needs.
                            @Override
                            public void onAuthenticationFailed() {}
                        });

        BiometricPrompt.PromptInfo.Builder info =
                new BiometricPrompt.PromptInfo.Builder()
                        .setTitle(title)
                        .setAllowedAuthenticators(AUTHENTICATORS);
        if (!subtitle.isEmpty()) {
            info.setSubtitle(subtitle);
        }

        try {
            activity.runOnUiThread(() -> prompt.authenticate(info.build()));
        } catch (Exception e) {
            call.resolve(result(false, "failed"));
        }
    }

    private JSObject result(boolean ok, String outcome) {
        JSObject out = new JSObject();
        out.put("ok", ok);
        out.put("outcome", outcome);
        return out;
    }
}

File: ios/App/App/Biometric/BiometricPlugin.swift
import Capacitor
import Foundation
import LocalAuthentication

/// The phone's own Face ID or Touch ID, asked for at the gate.
///
/// This proves the phone is in the hands of the person who set it up — a
/// different question from the one the selfie answers, which is whether the
/// face at the boundary belongs to the enrolled worker. It is asked first
/// because it is instant, and because a phone handed to a mate fails it
/// before a camera is ever opened.
///
/// It never replaces the selfie. A biometric says yes or no and leaves
/// nothing behind; attendance needs a record somebody can look at later.
///
/// `deviceOwnerAuthentication` rather than `...WithBiometrics`, so a passcode
/// is accepted when Face ID will not read. A site is dust, glare and a helmet
/// pulled low, and a worker locked out of their own day would simply stop
/// using the app. The passcode is the same proof of possession, slower.
@objc(BiometricPlugin)
public class BiometricPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "BiometricPlugin"
    public let jsName = "Biometric"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "verify", returnType: CAPPluginReturnPromise),
    ]

    /// What this device can do, asked before anything is shown.
    ///
    /// The web layer uses this to decide whether to put a step in front of
    /// check-in at all. A phone with nothing enrolled must not be given a
    /// gate it can never pass.
    @objc func isAvailable(_ call: CAPPluginCall) {
        let context = LAContext()
        var error: NSError?
        let ok = context.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error)

        var isFace = false
        if ok {
            // Reading biometryType before canEvaluatePolicy returns .none, so
            // the order here matters.
            isFace = context.biometryType == .faceID
        }

        call.resolve([
            "available": ok,
            "face": isFace,
            "reason": ok ? "ready" : Self.describe(error),
        ])
    }

    private static func describe(_ error: NSError?) -> String {
        guard let code = error.map({ LAError.Code(rawValue: $0.code) }) ?? nil else {
            return "unavailable"
        }
        switch code {
        case .biometryNotEnrolled, .passcodeNotSet: return "nothing enrolled"
        case .biometryNotAvailable: return "no hardware"
        case .biometryLockout: return "hardware busy"
        default: return "unavailable"
        }
    }

    /// Ask, and answer with what happened rather than only whether it worked.
    ///
    /// The caller has to tell a refusal from a sensor that cannot read:
    /// somebody who cancelled can try again, and somebody whose Face ID is
    /// locked out should be let through to the selfie rather than held at a
    /// door that will not open.
    @objc func verify(_ call: CAPPluginCall) {
        let reason = call.getString("subtitle").flatMap { $0.isEmpty ? nil : $0 }
            ?? call.getString("title")
            ?? "Confirm it's you"

        let context = LAContext()
        // The system sheet's fallback button. Named for what it does here
        // rather than left as "Enter Password", which reads like a different
        // account's password on a screen about a shift.
        context.localizedFallbackTitle = "Use passcode"

        var error: NSError?
        guard context.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error) else {
            call.resolve(["ok": false, "outcome": "unavailable"])
            return
        }

        context.evaluatePolicy(.deviceOwnerAuthentication, localizedReason: reason) { ok, err in
            var outcome = "ok"
            if !ok {
                let code = (err as NSError?).map { LAError.Code(rawValue: $0.code) } ?? nil
                switch code {
                case .userCancel, .appCancel, .systemCancel:
                    outcome = "cancelled"
                case .biometryNotAvailable, .biometryNotEnrolled, .passcodeNotSet:
                    outcome = "unavailable"
                default:
                    outcome = "failed"
                }
            }
            // Back to the main thread: the caller resumes a React render.
            DispatchQueue.main.async {
                call.resolve(["ok": ok, "outcome": outcome])
            }
        }
    }
}
